lib/spack/docs/pipelines.rst

```diff
@@ -0,0 +1,440 @@
+``install_missing_compilers``, so that if spack is asked to install a package
+with a compiler it doesn't know about, it can be quickly installed from the
+binary mirror first.
+
+Since bootstrapping compilers is optional, those items can be left out of the
+environment/stack file, and in that case no bootstrapping will be done (only the
+specs will be staged for building) and the runners will be expected to already
+have all needed compilers installed and configured for spack to use.
+
+-------------------------------------
+Using a custom spack in your pipeline
+-------------------------------------
+
+If your runners will not have a version of spack ready to invoke, or if for some
+other reason you want to use a custom version of spack to run your pipelines,
+this can be accomplished fairly simply.  First, your simple pipeline job needs
+to be augmented a bit compared to the very simple one provided at the beginning
+of this document.  Here's an example:
+
+.. code-block:: yaml
+
+   pipeline-job:
+     tags:
+       - <some-other-tag>
+   before_script:
+     - export SPACK_CLONE_LOCATION=$(mktemp -d)
+     - pushd ${SPACK_CLONE_LOCATION}
+     - git clone ${SPACK_REPO} --branch ${SPACK_REF}
+     - popd
+     - . ${SPACK_CLONE_LOCATION}/spack/share/spack/setup-env.sh
+   script:
+     - spack ci start <args>
+   after_script:
+     - rm -rf ${SPACK_CLONE_LOCATION}
+
+The environment variables ``SPACK_REPO`` and ``SPACK_REF`` are special, they are
+also described in the :ref:`ci_environment_variables` section.  Those environment
+variables are used to define a spack repository and branch/tag to use in running
+the pipeline.  If the ``spack ci start`` command sees those environment variables,
+then it adds similar ``before_script`` and ``after_script`` sections for each of
+the ``spack ci rebuild`` jobs which it generates.  This ensures that both the
+generation of the ``.gitlab-ci.yml`` and the conditional rebuilding of individual
+packages is done using the same custom version of spack.
+
+.. _ci_environment_variables:
+
+--------------------------------------------------
+Environment variables affecting pipeline operation
+--------------------------------------------------
+
+Certain secrets and some other information should be provided to the pipeline
+infrastructure via environment variables, usually for reasons of security, but
+in some cases to support other pipeline use cases such as PR testing.  The
+environment variables used by the pipeline infrastructure are described here.
+
+^^^^^^^^^^^^^^^^^
+AWS_ACCESS_KEY_ID
+^^^^^^^^^^^^^^^^^
+
+Needed when binary mirror is an S3 bucket.
+
+^^^^^^^^^^^^^^^^^^^^^
+AWS_SECRET_ACCESS_KEY
+^^^^^^^^^^^^^^^^^^^^^
+
+Needed when binary mirror is an S3 bucket.
+
+^^^^^^^^^^^^^^^
+S3_ENDPOINT_URL
+^^^^^^^^^^^^^^^
+
+Needed when binary mirror is an S3 bucket which is *not* on AWS.
+
+^^^^^^^^^^^^^^^^^
+CDASH_AUTH_TOKEN
+^^^^^^^^^^^^^^^^^
+
+Needed in order to report build groups to CDash.
```

Comment:
this should probably be provided as a secret and taken from the environment.  I wouldn't allow it in the YAML config